### PR TITLE
Add Co-Op Support

### PR DESCRIPTION
**Pull request by Dudejoe870**
This should add support for multiple controller emulation + Multiple Controllers at the same time + also with Keyboards also acting as different players. All Configurable in the Config, all the documentation should be updated too.

### CODE REVIEW

**Comment by Dudejoe870** [comment]
Uh oh, so it seems no homebrew input works with this PR, I'll start debugging it right now.

**Comment by Dudejoe870** [comment]
Whoops, I fixed the Merge Conflicts wrong, fixing it now.

File: Ryujinx.HLE/Hid/Hid.cs [hunk 1]
```diff
@@ -70,23 +75,132 @@ public Hid(Switch Device, long HidPosition)
             this.HidPosition = HidPosition;
 
             Device.Memory.FillWithZeros(HidPosition, Horizon.HidSize);
+			
+			Init();
+        }
+
+        private void Init()
+        {
+            if (HidEmulatedDevices.Devices.Handheld != -2)
```
Comment (ReinUsesLisp) [inline Ryujinx.HLE/Hid/Hid.cs]: Make Devices an enumerated array, something like this:
```
public enum DeviceId
{
    Handheld,
    Player1,
    Player2,
    Count
}
// Then
Devices = new WhateverTypeItIs[(int)DeviceId.Count];
```
You should write an DeviceId -> HidControllerId function, that way you can iterate through emulated devices without writting an if for every type.

File: Ryujinx.HLE/Hid/Hid.cs [hunk 2]
```diff
@@ -70,23 +75,132 @@ public Hid(Switch Device, long HidPosition)
             this.HidPosition = HidPosition;
 
             Device.Memory.FillWithZeros(HidPosition, Horizon.HidSize);
+			
+			Init();
```
Comment (ReinUsesLisp) [inline Ryujinx.HLE/Hid/Hid.cs]: It probably looks fine on your editor, there are some tabs here.

File: Ryujinx.HLE/Hid/Hid.cs [hunk 3]
```diff
@@ -70,23 +75,132 @@ public Hid(Switch Device, long HidPosition)
...
+                InitializeJoyconPair(
+                HidControllerId.CONTROLLER_HANDHELD,
+                HidControllerType.ControllerType_Handheld,
+                JoyConColor.Body_Neon_Red,
+                JoyConColor.Buttons_Neon_Red,
+                JoyConColor.Body_Neon_Blue,
+                JoyConColor.Buttons_Neon_Blue);
+            }
+
+            if (HidEmulatedDevices.Devices.Player1 != -2)
```
Comment (ReinUsesLisp) [inline Ryujinx.HLE/Hid/Hid.cs]: What does -2 mean? Add some enums or constants to avoid magic numbers.

File: Ryujinx.HLE/Hid/Hid.cs [hunk 4]
```diff
@@ -70,23 +75,132 @@ public Hid(Switch Device, long HidPosition)
...
+        {
+            if (HidEmulatedDevices.Devices.Handheld != -2)
+            {
+                InitializeJoyconPair(
+                HidControllerId.CONTROLLER_HANDHELD,
+                HidControllerType.ControllerType_Handheld,
+                JoyConColor.Body_Neon_Red,
+                JoyConColor.Buttons_Neon_Red,
+                JoyConColor.Body_Neon_Blue,
+                JoyConColor.Buttons_Neon_Blue);
```
Comment (ReinUsesLisp) [inline Ryujinx.HLE/Hid/Hid.cs]: Just a small change here, define a constant so JoyConColor can be easily changed (for future implementation).

File: Ryujinx/Config.cs [hunk 5]
```diff
@@ -12,8 +13,9 @@ namespace Ryujinx
 {
     public static class Config
     {
-        public static JoyConKeyboard   JoyConKeyboard   { get; private set; }
-        public static JoyConController JoyConController { get; private set; }
+        public static JoyConKeyboard     JoyConKeyboard        { get; private set; }
+        public static JoyConController[] JoyConControllers     { get; private set; }
+        public static bool               GamePadEnable         { get; private set; }
```
Comment (ReinUsesLisp) [inline Ryujinx/Config.cs]: There are some unneeded spaces here

File: Ryujinx/Config.cs [hunk 6]
```diff
@@ -157,6 +186,20 @@ private static ControllerInputID ToID(string Key)
...
             }
         }
+
+        // -2: None, -1: Keyboard, Everything Else: GamePad Index
+        private static int ToEmulatedDevice(string Key)
+        {
+            switch (Key.ToUpper())
+            {
+                case "NONE":     return -2;
+                case "KEYBOARD": return -1;
```
Comment (ReinUsesLisp) [inline Ryujinx/Config.cs]: Same as above, change these with an enum value or constant

File: Ryujinx/Ui/GLScreen.cs [hunk 7]
```diff
@@ -251,375 +252,42 @@ private new void UpdateFrame()
...
-                            LeftJoystickGamePad[HidEmulatedDevices.Devices.Player6],
-                            RightJoystickGamePad[HidEmulatedDevices.Devices.Player6]);
-                        break;
-                }
-            }
-
-            if (HidEmulatedDevices.Devices.Player7 != -2)
+        }
+        
+        private int GetGamePadIndexFromHostDevice(HidEmulatedDevices.HostDevice hostDevice)
```
Comment (ReinUsesLisp) [inline Ryujinx/Ui/GLScreen.cs]: nit: `hostDevice` -> `HostDevice`

File: Ryujinx/Ui/GLScreen.cs [hunk 8]
```diff
@@ -251,375 +252,42 @@ private new void UpdateFrame()
...
-                            HidControllerId.CONTROLLER_PLAYER_1,
-                            HidControllerLayouts.Main,
-                            CurrentButtonsGamePad[HidEmulatedDevices.Devices.Player1],
-                            LeftJoystickGamePad[HidEmulatedDevices.Devices.Player1],
-                            RightJoystickGamePad[HidEmulatedDevices.Devices.Player1]);
-                        break;
+                    Device.Hid.SetJoyconButton(
+                        entry.Key,
+                        (entry.Key == HidControllerId.CONTROLLER_HANDHELD) ? HidControllerLayouts.Handheld_Joined : HidControllerLayouts.Joined,
+                        (entry.Value == HidEmulatedDevices.HostDevice.Keyboard) ? CurrentButtonsKeyboard : CurrentButtonsGamePad[GetGamePadIndexFromHostDevice(entry.Value)],
```
Comment (ReinUsesLisp) [inline Ryujinx/Ui/GLScreen.cs]: Assign a boolean to avoid repeating this comparison.
You should also set these values into a local variable and then pass them to `Device.Hid.SetJoyconButton`, to avoid writting ternary operators in arguments (one is ok, but not for every argument).

File: Ryujinx/Ui/GLScreen.cs [hunk 9]
```diff
@@ -251,375 +252,42 @@ private new void UpdateFrame()
...
+                        (entry.Value == HidEmulatedDevices.HostDevice.Keyboard) ? LeftJoystickKeyboard   : LeftJoystickGamePad  [GetGamePadIndexFromHostDevice(entry.Value)],
+                        (entry.Value == HidEmulatedDevices.HostDevice.Keyboard) ? RightJoystickKeyboard  : RightJoystickGamePad [GetGamePadIndexFromHostDevice(entry.Value)]);
+                    Device.Hid.SetJoyconButton(
```
Comment (ReinUsesLisp) [inline Ryujinx/Ui/GLScreen.cs]: Space this one

File: Ryujinx/Ui/GLScreen.cs [hunk 10]
```diff
@@ -251,375 +252,42 @@ private new void UpdateFrame()
...
-                                HidControllerId.CONTROLLER_UNKNOWN,
-                                HidControllerLayouts.Main,
-                                CurrentButtonsGamePad[HidEmulatedDevices.Devices.PlayerUnknown],
-                                LeftJoystickGamePad[HidEmulatedDevices.Devices.PlayerUnknown],
-                                RightJoystickGamePad[HidEmulatedDevices.Devices.PlayerUnknown]);
-                            break;
-                    }
-                }
-            }
+            return -1;
```
Comment (ReinUsesLisp) [inline Ryujinx/Ui/GLScreen.cs]: nit: It'd be better to throw an ArgumentException if it's the argument is invalid. Returning `-1` will crash on unrelated code.

File: Ryujinx/Ui/GLScreen.cs (hunk 9 above)
Comment (Dudejoe870) [reply Ryujinx/Ui/GLScreen.cs]: I'm not exactly sure what you mean, can you explain a little bit?

Comment (ReinUsesLisp) [reply Ryujinx/Ui/GLScreen.cs]: Change:
```
    (entry.Value == HidEmulated ...
Device.Hid.SetJoyconButton
```
To:
```
    (entry.Value == HidEmulated ...

Device.Hid.SetJoyconButton
```

Comment (emmauss) [reply Ryujinx/Ui/GLScreen.cs]: nothing changed there

Comment (ReinUsesLisp) [reply Ryujinx/Ui/GLScreen.cs]: Mind the gap

Comment (firewolfgames) [reply Ryujinx/Ui/GLScreen.cs]: He added an empty line between those 2 lines @emmauss

File: Ryujinx/Ui/GLScreen.cs [hunk 11]
```diff
@@ -263,6 +263,7 @@ private new void UpdateFrame()
                         IsKeyboard ? CurrentButtonsKeyboard : CurrentButtonsGamePad[GetGamePadIndexFromHostDevice(entry.Value)],
                         IsKeyboard ? LeftJoystickKeyboard   : LeftJoystickGamePad  [GetGamePadIndexFromHostDevice(entry.Value)],
```
Comment (ReinUsesLisp) [inline Ryujinx/Ui/GLScreen.cs]: Just noticed it, `entry` -> `Entry`

File: Ryujinx.HLE/Hid/HidEmulatedDevices.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Ryujinx.HLE.Input
+{
+    public class HidEmulatedDevices
```
Comment (gdkchan) [inline Ryujinx.HLE/Hid/HidEmulatedDevices.cs]: I don't see why you need that. You can just pass a enum to InitializeJoycons, with a bit set for each controller that should be initialized/enabled.

**Comment by Dudejoe870** [comment]
Okay, I think I addressed most of everything.

**Comment by Dudejoe870** [comment]
Is anyone gonna give me a update on this?  Perhaps another review?

**Comment by DrHacknik** [comment]
I don't see any issues or conflicts; so I'm assuming it's safe to merge?

File: Ryujinx/Config.cs [hunk 13]
```diff
@@ -1,19 +1,23 @@
...
 using Ryujinx.UI.Input;
+using Ryujinx.HLE.Logging;
 using System;
 using System.Globalization;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
 using System.Reflection;
+using System.Text.RegularExpressions;
+using Ryujinx.Ui.Input;
```
Comment (emmauss) [inline Ryujinx/Config.cs]: sort usings

**Comment by emmauss** [comment]
Please rebase @Dudejoe870 . @gdkchan  are your reviews resolved?

**Comment by Dudejoe870** [comment]
Fixed the merge conflicts, I ran it and it seems fine.

**Comment by Dudejoe870** [comment]
Fixed the recent merge conflicts.

File: README.md [hunk 14]
```diff
@@ -95,9 +96,10 @@ Run `dotnet run -c Release -- path\to\game_exefs_and_romfs_folder` to run offici
...
 **Compatibility**
 
-You can check out the compatibility list within the Wiki. Only a handful of games actually work.
+You can check out the compatibility list within the Wiki.
 
 **Latest build**
 
 These builds are compiled automatically for each commit on the master branch. They may be unstable or might not work at all.  
 The latest automatic build for Windows (64-bit) can be found on the [official website](https://ryujinx.org/#/Build).
+However, for opening issues and testing, it is recommended to build the Master Branch as it is much more stable (Though still very experimental)
```
[the remark this answers is not in the document]
Comment (Dudejoe870) [reply README.md]: I wanted to finally fix the Readme, since no one has bothered to update it.

File: Ryujinx/Ui/GLScreen.cs [hunk 15]
```diff
@@ -128,51 +130,68 @@ public void MainLoop()
...
 
-            LeftJoystick = new HidJoystickPosition
+                    RightJoystickGamePad[i] = new HidJoystickPosition
+                    {
+                        DX = RightJoystickDXGamePad[i],
+                        DY = RightJoystickDYGamePad[i]
+                    };
+                }
+            }
+
```
[the remark this answers is not in the document]
Comment (Dudejoe870) [reply Ryujinx/Ui/GLScreen.cs]: I don't see any problems with spaces here, could you elaborate?

[the remark this answers is not in the document]
Comment (ReinUsesLisp) [reply Ryujinx/Ui/GLScreen.cs]: Thog means these spaces:
imagen

[the remark this answers is not in the document]
Comment (Dudejoe870) [reply Ryujinx/Ui/GLScreen.cs]: Ahhhh, sorry, it doesn't translate well on Github, I'll go ahead and fix it.

File: Ryujinx/Ui/Program.cs [hunk 16]
```diff
@@ -21,7 +21,9 @@ static void Main(string[] args)
             Switch Device = new Switch(Renderer, AudioOut);
 
             Config.Read(Device);
-
+
```
[the remark this answers is not in the document]
Comment (ReinUsesLisp) [reply Ryujinx/Ui/Program.cs]: @Dudejoe870 you forgot about this one

File: Ryujinx/Ui/Program.cs [hunk 17]
```diff
@@ -21,7 +21,9 @@ static void Main(string[] args)
             Switch Device = new Switch(Renderer, AudioOut);
 
             Config.Read(Device);
-
+          
+            Device.Hid.InitializeJoycons();
+
```
[the remark this answers is not in the document]
Comment (ReinUsesLisp) [reply Ryujinx/Ui/Program.cs]: and this one

File: README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (AcK77) [reply README.md]: Yeah, and I want to fix the Readme too... But another PR is required for that!

[the remark this answers is not in the document]
Comment (Dudejoe870) [reply README.md]: I don't see why, I just updated some of the outdated information, doesn't require a entirely different PR.

[the remark this answers is not in the document]
Comment (gdkchan) [reply README.md]: The build on the website is already a build from master, so I don't think that the users needs to compile it themselves just to test the emulator.

[the remark this answers is not in the document]
Comment (Dudejoe870) [reply README.md]: Yeah, I can update that, I think this was before you updated it to build the Master and not every single PR.

**Comment by Dudejoe870** [comment]
Closing this PR as no one probably cares about it at this point.
